tidb-server/main.go

```diff
@@ -403,6 +404,13 @@ func setGlobalVars() {
 	plannercore.SetPreparedPlanCache(config.CheckTableBeforeDrop || cfg.PreparedPlanCache.Enabled)
 	if plannercore.PreparedPlanCacheEnabled() {
 		plannercore.PreparedPlanCacheCapacity = cfg.PreparedPlanCache.Capacity
+		plannercore.PreparedPlanCacheMemoryGuardRatio = cfg.PreparedPlanCache.MemoryGuardRatio
+		plannercore.PreparedPlanCacheMaxMemory = cfg.Performance.MaxMemory
+		if plannercore.PreparedPlanCacheMaxMemory == 0 {
+			v, err := mem.VirtualMemory()
```

Comment:
According to https://github.com/shirou/gopsutil/blob/3b882b034ca24606010516bb521239cfcaf69cbd/mem/mem_linux.go#L40, though the function name is `VirtualMemory()`, `v.Total` actually is the physical memory of the machine, because the data source is `MemTotal` of `/proc/meminfo`.

While `ReadMemStats` returns golang memory allocator statistics, so it should be virtual memory consumption of this process(see also https://github.com/golang/go/blob/ae65615fd8784919f11e744b3a26d9dfa844c222/src/runtime/mstats.go#L573), so the check would be too strict indeed.

I still think it is pretty tricky to do memory accounting, it is kind of difficult for the DBA to set a proper value for this parameter, especially in cases where multiple TiDB servers are deployed in a single machine.

Reply:
@eurekaka 
Thank you for the detailed comments. 

The updated PR computes the total memory and the used one from the physical memory only. Thus, the PR should resolve the semantic mismatch in the previous PR.  

I agree that setting a proper value is not simple. However,  we should have provided a way to automatically calculate the value or allow the DBA to define the value in order to reduce or prevent the chance of OOM.